---
title: "VCF Quality Report"
date: "`r Sys.Date()`"
output: html_document
params:
  dat: ""
  chrom: ""
  wdirpop: ""
  bpen: ""
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, include=FALSE}
require(rmarkdown)
require(vcfR)
require(adegenet)
require(poppr)
library(yaml)
```


## Automatic VCF quality report

This document is a file automatically computed from a VCF candidate population dataset. It will help i the assessment of a dataset of sufficient quality for further analyses in polymorphism and LDhat pipelines.


```{r data, echo = TRUE}
dataset = params$dat
chrom = params$chrom
wdirpop = params$wdirpop
bpen = params$bpen

yaml_config = read_yaml(paste0("../../", wdirpop, "/", dataset, ".", chrom, ".bpen", bpen, ".yaml"))
K = yaml_config$K
pop = yaml_config$pop

vcf_file = paste("../../", wdirpop, "/", dataset, ".chromosome.", chrom, ".ldhat.vcf.gz", sep = "")
cat("Processing dataset", dataset, "chromosome", chrom, "\n")
print(vcf_file)

# Read the vcf
vcf = read.vcfR(vcf_file, verbose = FALSE)

#==================================================
# Size of the dataset
#==================================================
cat("Size of the dataset:")
utils:::format.object_size(file.info(vcf_file)$size, "auto")
```

  <!-- Summary stats of dataset - Automatic Rmd report -->
  <!--      Number of markers -->
  <!--      Number of individuals -->
  <!--      Dist of number of markers (non-missing sites) per individual -->
  <!--      Missing sites -->
  <!--      Hist of missing sites distribution -->
  <!--      Number of sites with missing sites < 10% -->
  <!--      Depth -->
  <!--      MAF -->
       

## Number of markers and individuals

```{r datasummary, echo = TRUE}
#==================================================
# Number of chromosomes and their names
#==================================================
length(unique(vcf@fix[,1]))
# Make the list of chromosomes
(list_chr = unique(vcf@fix[,1]))

#==================================================
# Number of positions
#==================================================
nrow(vcf@gt)

#==================================================
# Number of individuals
#==================================================
ncol(vcf@gt)
```

## VCF QUALITY ANALYSES

```{r genlight, echo = TRUE}
# Convert to genlight for faster computation and memory efficiency
# Large genomic dataset
genlight = vcfR2genlight(vcf)
genlight
```


### Missing data

```{r missing data, include = FALSE}
# Estimating number of missing data for each locus
missing = glNA(genlight)

# Distribution of proportions of missing data per site
hist(missing/nLoc(genlight), breaks = 40,
     main = "Proportions of missing data per site",
     xlab = "Missing data")
```


### Alleles

<!-- Number of alleles per sites and individuals. -->

<!-- ```{r echo = TRUE} -->
<!-- # Sum of second allele in each SNP -->
<!-- sumallele = glSum(genlight) -->
<!-- # Distribution of Minor Allele Frequencies -->
<!-- hist((sumallele/nLoc(genlight)), breaks = 40, -->
<!--      main = "Minor Allelic Frequencies", -->
<!--      xlab = "Minor Allelic Frequency") -->

<!-- # Mean number of second allele in each SNP -->
<!-- meanallele = glMean(genlight) -->

<!-- # Heatmap of the number of alleles per loci as a function of individuals -->
<!-- glPlot(genlight) -->
<!-- ``` -->


Minor Allele Frequency.

```{r include = FALSE}
# Element = 2 to target the minor allele
maf = vcfR::maf(vcf, element = 2)
```


```{r include = FALSE}
# Distribution of the minor allele frequency
hist(maf[,4], breaks = 100,
     main = "Minor Allelic Frequencies",
     xlab = "Minor Allelic Frequency")
abline(v = 0.005, col = "Red", lty = 2)
abline(v = 0.05, col = "Red", lty = 1)
```

Distribution of allele frequencies.

```{r include = FALSE}
AllFreq = glMean(genlight)
AllFreq = c(AllFreq, 1-AllFreq)
```

```{r include = FALSE}
hist(AllFreq, proba=TRUE, col="darkseagreen3", xlab="Allele frequencies",
main="Distribution of allele frequencies", nclass=20)
temp = density(AllFreq, bw=.05)
lines(temp$x, temp$y*2,lwd=3)
```

Number of SNPs without variant (i.e. only one allele).

```{r include = FALSE}
# Sum of second allele in each SNP
sumallele = glSum(genlight)
sum(sumallele==0)
```


Using chromR (vcfR) to locate unusual features in a genome, when quality information is available.

```{r chromosomes, echo = TRUE}
cat("===================================\n",
    "        Analysing chromosome", chrom, "\n",
    "===================================\n")
# Susbet for the chromosome of interest
vcf_subset=vcf[which(vcf@fix[,1] == chrom)]
# Create a chromR object.
chromosome = create.chromR(name = chrom, vcf = vcf_subset)
# Summary of the chromosome
chromosome

# Distribution of SNPs in the genome
hist(as.numeric(chromosome@vcf@fix[,2])/10^6, breaks = 1000,
     main = "Distribution of SNPs in the genome",
     xlab = "Position (Mb)")

# Distribution of missing data in the genome
# Estimate the proportion of missing data per individual
# genotypes in rows and individuals in columns
# For each row, how many sites are NA
NAsum = rowSums((is.na(chromosome@vcf@gt) | (chromosome@vcf@gt == ".") | (chromosome@vcf@gt == "./.")),
                na.rm = TRUE)

if (sum(!is.na(NAsum)) == 0) { # If only NA values
  # Plot proportion of missing data per SNP location
  plot(x = as.numeric(chromosome@vcf@fix[,2])/10^6, y = NAsum/nrow(chromosome@vcf@gt),
     main = "Distribution of missing data per site in the genome",
     xlab = "Position (Mb)", ylab = "Proportion of missing data",
     col = rgb(0, 0, 0, 0.1))
}

# Diagnostic plots (if information encoded in the vcf)
chromosome = proc.chromR(chromosome, verbose = TRUE)
plot(chromosome)
# Longitudinal analysis
chromoqc(chromosome, dp.alpha = 66)
# Tabular summary
head(chromosome@var.info)
head(chromosome@win.info)
```



